### PR TITLE
Jeanne Part 2

### PR DESCRIPTION
**Pull request by jeannefukumaru**

### CODE REVIEW

File: part2/2.2_flow-principles.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+2. Create simple failure reports using tools like MLflow or Jupyter notebooks with visualizations. ([methods](**Husain, H., & Shankar, S. (n.d.). Frequently Asked Questions (And Answers) about AI Evals – Hamel’s blog. Hamel’s Blog. https://hamel.dev/blog/posts/evals-faq/**))
+
+3. Gradually calibrate automated LLM judges metrics through MLflow scorers, 
+
+4. Translate these offline scorers to run in an online setting on production traces. Usually, this involves collecting traces into an OLAP database such as Delta tables, and then running batch jobs to evaluate a sample of these production traces using the same MLflow scorers.
+
+5. Surface aggregated metrics in dashboards to monitor trends over time. Periodically have product managers / business SMEs review and refine test suites based on new failure modes observed in production.
+
+### Proactively Manage Evaluation Costs
```
Comment (kwulffert23) [inline part2/2.2_flow-principles.md:27]: We could mention that we have "One time development cost" and "Recurring production monitoring cost" and the main variables driving costs there (dev one off: evaluators, number of eval dataset samples, iterations, tokens, prod: evaluators, percentage of traffic and traffic)

File: part2/2.2_flow-principles.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+3. Separate evaluation datasets into distinct classes oftest suites.
+CI tests can be run quickly by limiting the number of input examples to a small number, for example approximately 100 examples. The main goal is to identify regressions before production deployment. 
+
+Production test suites can run on a larger sample of live traces through an offline batch job. The goal here is different to that of CI testing. Here, we are less time-constrained, and we want a large enough sample to allow us to identify new, uncategorized errors for triage.
+
+Tuning Automated LLM Judges for Reliability
+
+LLM judges require careful calibration to ensure consistent, reliable evaluations across different scenarios and edge cases.
+
+### Enabling Easy Rollbacks During Failures
```
Comment (kwulffert23) [inline part2/2.2_flow-principles.md:51]: Nicely written and crisp!

File: part2/2.3_feedback-principles.qmd
Comment (kwulffert23) [inline part2/2.3_feedback-principles.qmd:1]: Looks great!

File: part2/2.4_continuous-learning.qmd [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+- These customers typically come to us and ask not only how to ensure the quality of a single POC or MVP, but how to build reliable best practices that allow an organisation to scale how the organisation applies generative AI to improve business practices. 
+- Scaling generative AI best practices across an organisation can be achieved in a few ways:
+    - institutionalizing technical best practices through standardised frameworks, reference architectures and best practices documentation
+    - investing in continuously refining how business and technical stakeholders work together to uncover how to use generative AI to accomplish their goals
+
+## Standardised frameworks
+Standarised frameworks allow us to achieve consistent, reliable, and efficient delivery at scale. An efficient framework aims to make software development and deployment feel like a well-oiled assembly line rather than a series of one-off custom projects, ultimately enabling faster time-to-market with higher quality and lower risk.
+
+An example of a standardised framework that can be used for generative AI applications is Databricks Asset Bundles (DAB)
```
Comment (kwulffert23) [inline part2/2.4_continuous-learning.qmd:23]: shall we add the link to the DABs repo?

**Review by kwulffert23 — COMMENTED** [review]
Hi @jeannefukumaru , great content. I left a couple of comments.

File: part2/2.2_flow-principles.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+[*"Our goal is to decrease the amount of time required for changes to be deployed into production and to increase the reliability and quality of those services."* ](**The DevOps Handbook: How to Create World-Class Agility, Reliability, & Security in Technology Organizations eBook : Kim bestselling author of The, Gene, Humble, Jez, Debois, Patrick, Willis, John, Forsgren, Nicole: Amazon.co.uk: Books. (n.d.). https://www.amazon.co.uk/DevOps-Handbook-World-Class-Reliability-Organizations-ebook/dp/B0F7J1WWQD?ref_=ast_author_mpb**)
+
+## Implementing Efficient Flows
+
+### Creating Quick Feedback Loops Through Systematic Test Suite Development
+Unit testing and integration testing in software engineering can usually be defined upfront because we have an idea of the behaviour of the system we want to build, including it's edge cases. Generative AI is different. Creating tests needs an additional preliminary step of  extensively analysing traces of previous requests and responses in order to identify common failures modes. Only when these failure modes are identified and categorised can be we then defining automated checks to identify those failure patterns.
```
[the remark this answers is not in the document]
Comment (jeannefukumaru) [reply part2/2.2_flow-principles.md]: Added a paragraph

"It's worth noting that seasoned traditional software developers often recommend "writing tests first". However, writing a full test suite upfront is rarely feasible with Generative AI. Unlike traditional software engineering, where we can reasonably anticipate the range of failure modes and edge cases, LLM outputs are less predictable in their outputs. Moreover, LLM capabilities differ across models and are constantly evolving with each new model release. Therefore, it's more often the case that edge cases have to be "discovered" and codified with more empirical methods."

File: part2/2.2_flow-principles.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+** Dangerous default** 
+
+Teams often fall into a dangerous default pattern where one tester runs five to six queries repeatedly in a chat interface and provides pass/fail analysis. This approach fails to capture error frequency metrics or classify failure types systematically.
+
+Teams should instead apply the scientific method iteratively. 
+
+1. Start by categorizing failure modes as quickly as possible using approximately 100 traces. ([source](**Husain, H., & Shankar, S. (n.d.). Frequently Asked Questions (And Answers) about AI Evals – Hamel’s blog. Hamel’s Blog. https://hamel.dev/blog/posts/evals-faq/**))
+
+2. Create simple failure reports using tools like MLflow or Jupyter notebooks with visualizations. ([methods](**Husain, H., & Shankar, S. (n.d.). Frequently Asked Questions (And Answers) about AI Evals – Hamel’s blog. Hamel’s Blog. https://hamel.dev/blog/posts/evals-faq/**))
```
[the remark this answers is not in the document]
Comment (jeannefukumaru) [reply part2/2.2_flow-principles.md]: Added some screenshot examples from line 23

File: part2/2.2_flow-principles.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
```
[the remark this answers is not in the document]
Comment (jeannefukumaru) [reply part2/2.2_flow-principles.md]: added "3. Gradually calibrate automated LLM judges metrics through MLflow scorers and cross-checking with domain experts"

File: part2/2.2_flow-principles.md (hunk 1 above)
Comment (jeannefukumaru) [reply part2/2.2_flow-principles.md]: Rephrased to "Unlike traditional unit tests, LLM-as-a-judge calls cost money. Agent traces can span significant lengths, increasing token costs per test. Additionally, production monitoring entails running evaluation suites on a recurring basis, which can increase costs even further. Therefore, teams need strategies for creating fast and affordable evaluation suites."

**Comment by jeannefukumaru** [comment]
Merging pull request after addressing comments. Changes to subsequent drafts can be separate pull requests
